In [1]:
import dask
import numpy as np
import sklearn
from matplotlib import pyplot as plt
from matplotlib import cm
import copy
import astropy
import hdbscan
import pandas as pd

from astropy.io import fits

In [2]:
# hdul = fits.open('http://gal-03.sai.msu.ru/~vtoptun/photometry/rcsed_v2_clean.fits', memmap=astropy.io.fits.Conf.use_memmap.defaultvalue, lazy_load_hdus=True)
# hdul = fits.open('http://gal-03.sai.msu.ru/~vtoptun/redshift/rcsed_v2.fits')
hdul = fits.open('rcsed_v2_clean.fits', memmap=astropy.io.fits.Conf.use_memmap.defaultvalue, lazy_load_hdus=True)

In [3]:
data = hdul[1].data

In [4]:
hdul.close()

In [5]:
DATA = pd.DataFrame(np.array(data).byteswap().newbyteorder())

In [6]:
del data

In [ ]:
# DATA.to_csv('rcsed_v2_clean.csv')
# del DATA

In [7]:
DATA.shape

(4105516, 559)

In [65]:
(~DATA.z_sdss.isna()).sum()

3169675

In [8]:
DATA.info(verbose=True,null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4105516 entries, 0 to 4105515
Data columns (total 559 columns):
ind                          4105516 non-null int32
ra                           4105516 non-null float64
dec                          4105516 non-null float64
z_sdss                       3169675 non-null float64
zerr_sdss                    3163704 non-null float64
specObjID_sdss               3171529 non-null float64
bestObjID_sdss               4105516 non-null int64
plate_sdss                   4105516 non-null int16
mjd_sdss                     4105516 non-null int32
fiberID_sdss                 4105516 non-null int16
z_2df                        220438 non-null float32
q_z_2df                      4105516 non-null int32
SeqNum_2df                   4105516 non-null object
z_6df                        116200 non-null float64
quality_6df                  4105516 non-null int16
targetname_6df               4105516 non-null object
specid_6df                   4105516 non

In [9]:
pd.set_option('max_columns', None)
DATA.head()

ind        ra        dec    z_sdss  zerr_sdss  specObjID_sdss  \
0    1  0.000019  -4.760800  2.467120   0.002194    7.919679e+18   
1    2  0.000091  24.902252  0.555936   0.000245    7.333218e+18   
2    3  0.000346  -6.491940  0.514177   0.000130    8.047035e+18   
3    4  0.000463  10.241074  0.654380   0.000170    6.960408e+18   
4    5  0.000559  34.985602  0.146326   0.000021    8.044630e+18   

        bestObjID_sdss  plate_sdss  mjd_sdss  fiberID_sdss  z_2df     q_z_2df  \
0                    0        7034     56564           360    NaN -2147483648   
1  1237666309095751941        6513     56543           843    NaN -2147483648   
2  1237680243597902184        7147     56574           831    NaN -2147483648   
3  1237679319089349467        6182     56190           346    NaN -2147483648   
4  1237666184574271705        7145     56567           274    NaN -2147483648   

  SeqNum_2df  z_6df  quality_6df      targetname_6df  specid_6df   recno_uzc  \
0  b'      '    NaN       -32768  b'               ' -2147483648 -2147483648   
1  b'      '    NaN       -32768  b'               ' -2147483648 -2147483648   
2  b'      '    NaN       -32768  b'               ' -2147483648 -2147483648   
3  b'      '    NaN       -32768  b'               ' -2147483648 -2147483648   
4  b'      '    NaN       -32768  b'               ' -2147483648 -2147483648   

   z_uzc  z_err_uzc  z_cfa  z_err_cfa  z_hectospec  z_lamost  z_err_lamost  \
0    NaN        NaN    NaN        NaN          NaN       NaN           NaN   
1    NaN        NaN    NaN        NaN          NaN       NaN           NaN   
2    NaN        NaN    NaN        NaN          NaN       NaN           NaN   
3    NaN        NaN    NaN        NaN          NaN       NaN           NaN   
4    NaN        NaN    NaN        NaN          NaN       NaN           NaN   

   obsid_lamost  lmjd_lamost  mjd_lamost  \
0   -2147483648  -2147483648 -2147483648   
1   -2147483648  -2147483648 -2147483648   
2   -2147483648  -2147483648 -2147483648   
3   -2147483648  -2147483648 -2147483648   
4   -2147483648  -2147483648 -2147483648   

                                 planid_lamost  spid_lamost  fiberid_lamost  \
0  b'                                        '          255             255   
1  b'                                        '          255             255   
2  b'                                        '          255             255   
3  b'                                        '          255             255   
4  b'                                        '          255             255   

   z_lega_c  f_z_lega_c  f_spec_lega_c SPECT_ID_lega_c  Z_deep2  OBJNO_deep2  \
0       NaN      -32768         -32768   b'          '      NaN  -2147483648   
1       NaN      -32768         -32768   b'          '      NaN  -2147483648   
2       NaN      -32768         -32768   b'          '      NaN  -2147483648   
3       NaN      -32768         -32768   b'          '      NaN  -2147483648   
4       NaN      -32768         -32768   b'          '      NaN  -2147483648   

   Z_deep3  OBJNO_deep3  z_wigglez  Q_wigglez   SpecFile_wigglez  Z_gama  \
0      NaN  -2147483648        NaN     -32768  b'              '     NaN   
1      NaN  -2147483648        NaN     -32768  b'              '     NaN   
2      NaN  -2147483648        NaN     -32768  b'              '     NaN   
3      NaN  -2147483648        NaN     -32768  b'              '     NaN   
4      NaN  -2147483648        NaN     -32768  b'              '     NaN   

   NQ_gama                           SPECID_gama  multiple_z  dz      E_bv  \
0   -32768  b'                                 '          70  70  0.032031   
1   -32768  b'                                 '          70  70  0.081351   
2   -32768  b'                                 '          70  70  0.037788   
3   -32768  b'                                 '          70  70  0.087513   
4   -32768  b'                                 '          70  70  0.077543   

   A_u_sds

**Наблюдения:**
- в некоторых колонках,  в которых присутствуют все значния есть очень странные значения вроде "-2147483648" или "-32768" и другие

In [83]:
#    Здесь мы создадим датафрейм с наиболее 
#       частотными значениями по каждому из столбцов

top_values = []

for col in DATA.columns[1:]:
    vc = DATA[col].value_counts(ascending=False).head(1)
    top_values.append([vc.name, vc.index[0], vc[vc.index[0]]])
    
anomaly = pd.DataFrame(data=top_values, columns = ['column_name', 'anomaly_value', 'value_counts'])

In [84]:
# Здесь мы создаем колонку с количеством
#  ненулевых значений в каждом из столбцов таблицы
#     DATA, начиная с ra

not_nan = []

for col in DATA.columns[1:]:
    not_nan.append((~DATA[col].isna()).sum())
    
not_nan = pd.Series(not_nan)

In [85]:
# Дополним получившийся датафрейм столбцом, показывающим,
# какую долю от длины всего столбца занимает каждое аномальное значение

import copy

top_values_data = copy.copy(anomaly)

top_values_data['percent_total'] = (top_values_data.value_counts / len(DATA)).round(3)
top_values_data['percent_not_nan'] = (top_values_data.value_counts / not_nan).round(3)

In [86]:
# Выведем на экран получившйся датафрейм

with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(top_values_data.sort_values(by='percent_not_nan', ascending=False))

column_name                                anomaly_value  \
18                   z_err_uzc                                            0   
33             SPECT_ID_lega_c                                b'          '   
32               f_spec_lega_c                                       -32768   
245        KSERMAG2DERR_ukidss                                 -9.99999e+08   
244           KSERMAG2D_ukidss                                 -9.99999e+08   
243          KPSFMAGERR_ukidss                                 -9.99999e+08   
242             KPSFMAG_ukidss                                 -9.99999e+08   
31                  f_z_lega_c                                       -32768   
20                   z_err_cfa                                            0   
37                 OBJNO_deep3                                  -2147483648   
16                   recno_uzc                                  -2147483648   
45                          dz                                           70   
514        phot_data_new_galex                                           84   
35                 OBJNO_deep2                                  -2147483648   
14              targetname_6df                           b'               '   
13                 quality_6df                                       -32768   
15                  specid_6df                                  -2147483648   
43                 SPECID_gama         b'                                 '   
42                     NQ_gama                                       -32768   
515              wigglez_photo                                           70   
44                  multiple_z                                           70   
11                  SeqNum_2df                                    b'      '   
10                     q_z_2df                                  -2147483648   
28                 spid_lamost                                          255   
39                   Q_wigglez                                       -32768   
40            SpecFile_wigglez                            b'              '   
29              fiberid_lamost                                          255   
25                 lmjd_lamost                                  -2147483648   
27               planid_lamost  b'                                        '   
26                  mjd_lamost                                  -2147483648   
24                obsid_lamost                                  -2147483648   
464            w4sigmag_1_wise                                         9999   
478            w4sigmag_8_wise                                         9999   
474            w4sigmag_6_wise                                         9999   
476            w4sigmag_7_wise                                         9999   
472            w4sigmag_5_wise                                         9999   
462              w4sigmag_wise                                         9999   
468            w4sigmag_3_wise                                         9999   
470            w4sigmag_4_wise                                         9999   
466            w4sigmag_2_wise                                         9999   
446            w3sigmag_1_wise                                         9999   
481      NUV_KRON_RADIUS_galex                                          3.5   
170          kron_radius_z_des                                          3.5   
169          kron_radius_i_des                                          3.5   
168          kron_radius_r_des                                          3.5   
171          kron_radius_y_des                                          3.5   
167          kron_radius_g_des                                          3.5   
257           yPetroMagErr_vhs                                 -9.99999e+08   
256              yPetroMag_vhs                                 -9.99999e+08   
260              yAperMag4_vhs                                 -9.999

### Анализ данных на основе таблицы top_values_data

Таблица **top_values_data** содержит в себе следующие столбцы:
- **column_name**   -   названия столбцов таблицы **DATA**

- **anomaly_value**   -   наиболее частотные значения столбцов таблицы **DATA**

- **value_counts**   -   количество раз, которое *anomaly_value* встречается в столбце таблицы **DATA**

- **percent_total**   -   доля *anomaly_value* от всего объема столбца таблицы **DATA**

- **percent_not_nan**   -   доля *anomaly_value* от non-null объема столбца таблицы **DATA**

В ячейке ниже показано, сколько столбцов таблицы **DATA** содержат более **90% и т.д.** аномальных значений от общего количества non-null значений столбца. Оказывается, что около трети столбцов (162) содержат таки аномалий больше, чем **10%**.

**! Нужно понять, какой порог считать допустимым.**

**! Остальные можно отбросить.**


In [113]:
for num in np.arange(0.9,0.0,-0.05).round(2):
    print(num, '   ', len(top_values_data.query('percent_not_nan > @num')))

0.9     32
0.85     40
0.8     40
0.75     40
0.7     42
0.65     71
0.6     71
0.55     88
0.5     88
0.45     93
0.4     101
0.35     101
0.3     103
0.25     108
0.2     114
0.15     117
0.1     162
0.05     219


In [116]:
#     Добавим столбец non_null_part, показываающий 
#     какую часть от величины столбца составляют ненулевые значения

top_values_data['non_null_part'] = (not_nan / len(DATA)).round(3)

with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(top_values_data.sort_values(by='non_null_part', ascending=True))

column_name                                anomaly_value  \
30                    z_lega_c                                       0.7282   
180          u_petro_skymapper                                       18.287   
182          v_petro_skymapper                                       17.637   
183        e_v_petro_skymapper                                        0.134   
181        e_u_petro_skymapper                                        0.147   
36                     Z_deep3                                     0.533413   
17                       z_uzc                                    0.0171952   
18                   z_err_uzc                                            0   
184          g_petro_skymapper                                       18.139   
34                     Z_deep2                                      1.24116   
185        e_g_petro_skymapper                                        0.147   
20                   z_err_cfa                                            0   
19                       z_cfa                                    0.0170118   
186          r_petro_skymapper                                       17.728   
187        e_r_petro_skymapper                                        0.134   
350                  Rpmag_vst                                      17.2979   
348                   Gap6_vst                                      18.6666   
347                 e_Gap4_vst                                       0.0084   
349                 e_Gap6_vst                                        0.007   
346                   Gap4_vst                                       19.126   
343                e_Gpmag_vst                                       0.0127   
342                  Gpmag_vst                                      18.9186   
341                 e_Uap6_vst                                       0.0673   
340                   Uap6_vst                                      19.9833   
339                 e_Uap4_vst                                       0.0747   
338                   Uap4_vst                                      20.5147   
337                 e_Uap3_vst                                       0.0901   
336                   Uap3_vst                                       20.677   
335                e_Upmag_vst                                       0.0906   
334                  Upmag_vst                                      20.1305   
344                   Gap3_vst                                      19.5856   
351                e_Rpmag_vst                                       0.0087   
345                 e_Gap3_vst                                       0.0091   
353                 e_Rap3_vst                                       0.0053   
373                 e_Zap6_vst                                       0.0099   
372                   Zap6_vst                                      17.2466   
371                 e_Zap4_vst                                       0.0091   
370                   Zap4_vst                                      17.7277   
369                 e_Zap3_vst                                       0.0112   
368                   Zap3_vst                                      17.6196   
367                e_Zpmag_vst                                       0.0179   
366                  Zpmag_vst                                       16.601   
365                 e_Iap6_vst                                       0.0054   
352                   Rap3_vst                                      18.4782   
364                   Iap6_vst                                      17.5468   
362                   Iap4_vst                                      17.9821   
361                 e_Iap3_vst                                       0.0058   
360                   Iap3_vst                                      18.8354   
359                e_Ipmag_vst                                       0.0089   
358                  Ipmag_vst                                      1

In [119]:
# Выясним сколько столбцов таблицы DATA  заполнены меньше чем на num*100 %

for num in np.arange(0.9,0.0,-0.1).round(2):
    print(num, '   ', len(top_values_data.query('non_null_part < @num')))

0.9     479
0.8     419
0.7     414
0.6     414
0.5     373
0.4     301
0.3     258
0.2     202
0.1     120


In [120]:
len(top_values_data.query('(non_null_part<0.01)&(percent_not_nan<0.5)'))

10

In [114]:
# Выведем на экран получившйся датафрейм

with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(top_values_data.sort_values(by='value_counts', ascending=False))

column_name                                anomaly_value  \
33             SPECT_ID_lega_c                                b'          '   
31                  f_z_lega_c                                       -32768   
32               f_spec_lega_c                                       -32768   
37                 OBJNO_deep3                                  -2147483648   
16                   recno_uzc                                  -2147483648   
45                          dz                                           70   
514        phot_data_new_galex                                           84   
35                 OBJNO_deep2                                  -2147483648   
15                  specid_6df                                  -2147483648   
14              targetname_6df                           b'               '   
13                 quality_6df                                       -32768   
43                 SPECID_gama         b'                                 '   
42                     NQ_gama                                       -32768   
515              wigglez_photo                                           70   
44                  multiple_z                                           70   
10                     q_z_2df                                  -2147483648   
11                  SeqNum_2df                                    b'      '   
40            SpecFile_wigglez                            b'              '   
39                   Q_wigglez                                       -32768   
24                obsid_lamost                                  -2147483648   
25                 lmjd_lamost                                  -2147483648   
26                  mjd_lamost                                  -2147483648   
27               planid_lamost  b'                                        '   
28                 spid_lamost                                          255   
29              fiberid_lamost                                          255   
464            w4sigmag_1_wise                                         9999   
476            w4sigmag_7_wise                                         9999   
478            w4sigmag_8_wise                                         9999   
468            w4sigmag_3_wise                                         9999   
470            w4sigmag_4_wise                                         9999   
474            w4sigmag_6_wise                                         9999   
472            w4sigmag_5_wise                                         9999   
466            w4sigmag_2_wise                                         9999   
462              w4sigmag_wise                                         9999   
446            w3sigmag_1_wise                                         9999   
448            w3sigmag_2_wise                                         9999   
450            w3sigmag_3_wise                                         9999   
452            w3sigmag_4_wise                                         9999   
454            w3sigmag_5_wise                                         9999   
444              w3sigmag_wise                                         9999   
456            w3sigmag_6_wise                                         9999   
458            w3sigmag_7_wise                                         9999   
460            w3sigmag_8_wise                                         9999   
481      NUV_KRON_RADIUS_galex                                          3.5   
243          KPSFMAGERR_ukidss                                 -9.99999e+08   
245        KSERMAG2DERR_ukidss                                 -9.99999e+08   
244           KSERMAG2D_ukidss                                 -9.99999e+08   
242             KPSFMAG_ukidss                                 -9.99999e+08   
389     gMeanKronMag_panstarrs                                         -999   
390  gMeanKronMagErr_panstarrs                                       

In [10]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(DATA.describe(include='object'))

SeqNum_2df      targetname_6df  \
count     4105516             4105516   
unique     222290              115740   
top     b'      '  b'               '   
freq      3882161             3989227   

                                      planid_lamost SPECT_ID_lega_c  \
count                                       4105516         4105516   
unique                                         3143            1796   
top     b'                                        '   b'          '   
freq                                        3880087         4103699   

         SpecFile_wigglez                           SPECID_gama  
count             4105516                               4105516  
unique             223883                                158365  
top     b'              '  b'                                 '  
freq              3881571                               3946981

In [11]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(DATA.describe())

ind            ra           dec        z_sdss     zerr_sdss  \
count  4.105516e+06  4.105516e+06  4.105516e+06  3.169675e+06  3.163704e+06   
mean   2.123240e+06  1.734465e+02  1.831309e+01  7.018208e-01  2.850560e-03   
std    1.224227e+06  9.467727e+01  2.347748e+01  8.312641e-01  2.072000e+00   
min    1.000000e+00  1.913209e-05 -8.933516e+01  1.600210e-03  5.358193e-09   
25%    1.067532e+06  1.293579e+02  1.249885e+00  1.942430e-01  3.473064e-05   
50%    2.125360e+06  1.766058e+02  1.749696e+01  4.909693e-01  1.191420e-04   
75%    3.179898e+06  2.232626e+02  3.576123e+01  6.813946e-01  2.175503e-04   
max    4.240803e+06  3.599999e+02  8.934769e+01  7.055639e+00  3.614933e+03   

       specObjID_sdss  bestObjID_sdss    plate_sdss      mjd_sdss  \
count    3.171529e+06    4.105516e+06  4.105516e+06  4.105516e+06   
mean     5.064625e+18   -1.186490e+18 -3.979712e+03 -4.885005e+08   
std      2.670082e+18    4.367370e+18  1.576099e+04  9.002810e+08   
min      2.994897e+17   -9.223372e+18 -3.276800e+04 -2.147484e+09   
25%      2.472540e+18    0.000000e+00  4.810000e+02  5.199300e+04   
50%      5.350320e+18    1.237661e+18  3.948000e+03  5.538300e+04   
75%      7.229571e+18    1.237666e+18  5.978000e+03  5.620900e+04   
max      1.125927e+19    1.237681e+18  1.000000e+04  5.752000e+04   

       fiberID_sdss          z_2df       q_z_2df          z_6df   quality_6df  \
count  4.105516e+06  220438.000000  4.105516e+06  116200.000000  4.105516e+06   
mean  -7.111565e+03       0.117832 -2.030653e+09       0.112909 -3.183974e+04   
std    1.392505e+04       0.066185  4.870761e+08       0.330948  5.436821e+03   
min   -3.276800e+04       0.001610 -2.147484e+09       0.001607 -3.276800e+04   
25%    2.700000e+01       0.076540 -2.147484e+09       0.036098 -3.276800e+04   
50%    2.990000e+02       0.111130 -2.147484e+09       0.055507 -3.276800e+04   
75%    5.770000e+02       0.151310 -2.147484e+09       0.079953 -3.276800e+04   
max    1.000000e+03       3.499890  5.000000e+00       5.981890  6.000000e+00   

         specid_6df     recno_uzc         z_uzc     z_err_uzc         z_cfa  \
count  4.105516e+06  4.105516e+06  19365.000000  1.937300e+04  40716.000000   
mean  -2.086654e+09 -2.137295e+09      0.024437  7.575914e-09      0.043737   
std    3.562785e+08  1.475702e+08      0.012735  1.054468e-06      0.049944   
min   -2.147484e+09 -2.147484e+09      0.001604  0.000000e+00      0.001601   
25%   -2.147484e+09 -2.147484e+09      0.016031  0.000000e+00      0.015868   
50%   -2.147484e+09 -2.147484e+09      0.023490  0.000000e+00      0.027702   
75%   -2.147484e+09 -2.147484e+09      0.031789  0.000000e+00      0.050698   
max    1.397370e+05  1.936700e+04      0.231977  1.467682e-04      0.333197   

          z_err_cfa    z_hectospec       z_lamost   z_err_lamost  \
count  4.072700e+04  159344.000000  181205.000000  225429.000000   
mean   1.392341e-09       0.339105       0.421040   -2571.653320   
std    2.809875e-07       0.219975       0.678689    4372.810547   
min    0.000000e+00       0.001601       0.001600   -9999.000000   
25%    0.000000e+00       0.180577       0.066273   -9999.000000   
50%    0.000000e+00       0.297015       0.110470       0.000018   
75%    0.000000e+00       0.447978       0.234780       0.000083   
max    5.670589e-05       2.113730       5.707270   59966.898438   

       obsid_lamost   lmjd_lamost    mjd_lamost   spid_lamost  fiberid_lamost  \
count  4.105516e+06  4.105516e+06  4.105516e+06  4.105516e+06    4.105516e+06   
mean  -2.011889e+09 -2.029565e+09 -2.029565e+09  2.414661e+02    2.479284e+02   
std    5.637659e+08  4.892145e+08  4.892145e+08  5.615880e+01    3.367907e+01   
min   -2.147484e+09 -2.147484e+09 -2.147484e+09  1.000000e+00    1.000000e+00   
25%   -2.147484e+09 -2.147484e+09 -2.147484e+09  2.550000e+02    2.550000e+02   
50%   -2.147484e+09 -2.147484e+09 -2.147484e+09  2.550000e+02    2.550000e+02   
75%   -2.147484e+09 -2.147484e+09 -2.147484e+09  2.550000e+0